org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/symbol/DocumentSymbolService.java

```diff
@@ -96,12 +100,12 @@ public class DocumentSymbolService implements IDocumentSymbolService {
 	public List<? extends Location> getDefinitions(Document document, XtextResource resource, DefinitionParams params,
 			IReferenceFinder.IResourceAccess resourceAccess, CancelIndicator cancelIndicator) {
 		int offset = document.getOffSet(params.getPosition());
-		return getDefinitions(resource, offset, resourceAccess, cancelIndicator);
+		EObject element = eObjectAtOffsetHelper.resolveCrossReferencedElementAt(resource, offset);
+		return getDefinitions(resource, element, resourceAccess, cancelIndicator);
 	}
 
-	public List<? extends Location> getDefinitions(XtextResource resource, int offset,
+	private List<? extends Location> getDefinitions(XtextResource resource, EObject element,
```

Comment:
Thank you for adding a test.

The more I look at this, the more I wonder how it should be if designed from scratch.

With this patch, we are using resolvedCrossReferenced when asking for definitions, but line 146 still uses the contained object at the cursor position. So asking for definitions will yield different results than asking for references when includeDefinition is set to true. I've doubts about it.

Also there appears to be quite some duplication at first or second glance.

I'd think it would make sense to extract the part that computes the targetURIs at a given location (offset) into an own method. Subsequently `definitions` and `references` would work with that collection. How does that sound to you?

Reply:
is there a list of usecases/examples for these methods?

Reply:
I do not know about the methods, I just copied them. The examples of what I would like to achieve are added to the test cases, for example, get definition on a comment (test case 3) or a keyword (test case 4) does not return a definition.

Reply:
how does this differ from what HyperlinkHelper does in ui?

Reply:
HyperlinkHelper in ui only create links for cross-references, as in the version of this PR which submit first. Since Sebastian was skeptical about this I looked for some other condition which still meets the requirements.

Reply:
@szarnekow how shall we continze here?

Reply:
Hi @cdietrich , @szarnekow , can I help to make this PR move forward?

Reply:
i am waiting for feedback from sebastian. we need to be a bit patient. anyway you should rebase as we have bumped the version to 2.28. this also needs changes to the sinces